### ISSUE TITLE
crouton: version 1-20150226113948~master:060367f9: fails to start with seg fault

### ISSUE THREAD

**Issue by Watney** [dnschneid/crouton#1443]
Output from ``croutonversion``
crouton: version 1-20150226113948~master:060367f9
release: trusty
architecture: amd64
xmethod: xorg
targets: xorg,xfce,keyboard,extension,xiwi
host: version 6812.3.0 (Official Build) dev-channel glimmer_freon 

``sudo startxfce4``
(EE) Backtrace:
(EE) 0: /usr/bin/Xorg (xorg_backtrace+0x48) [0x7f212faaf848]
(EE) 1: /usr/bin/Xorg (0x7f212f906000+0x1ad539) [0x7f212fab3539]
(EE) 2: /lib/x86_64-linux-gnu/libpthread.so.0 (0x7f212e800000+0x10340) [0x7f212e810340]
(EE) 3: /usr/bin/Xorg (0x7f212f906000+0xb57a6) [0x7f212f9bb7a6]
(EE) 4: /usr/bin/Xorg (xf86BusProbe+0x9) [0x7f212f98f099]
(EE) 5: /usr/bin/Xorg (InitOutput+0x74d) [0x7f212f99d6fd]
(EE) 6: /usr/bin/Xorg (0x7f212f906000+0x59bab) [0x7f212f95fbab]
(EE) 7: /lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main+0xf5) [0x7f212d240ec5]
(EE) 8: /usr/bin/Xorg (0x7f212f906000+0x451ee) [0x7f212f94b1ee]
(EE) 
(EE) Segmentation fault at address 0x0
(EE) Fatal server error:
(EE) Caught signal 11 (Segmentation fault). Server aborting
(EE) 
(EE) 
Please consult the The X.Org Foundation support  at http://wiki.x.org for help. 
(EE) Please also check the log file at "/tmp/Xorg.16689.log" for additional information.
(EE) 
(EE) Server terminated with error (1). Closing log file.``

I also attempt to start with the xiwi target, and it throws the same error

**Comment by dnschneid** [comment]
Could you please try the following:
```
sudo enter-chroot -u root sed -i '/AutoAddDevices/a/Option "AutoAddGPU" "false"/' /etc/crouton/xorg-dummy.conf
```
and try xiwi again?

**Comment by Watney** [comment]
++) Log file: "/tmp/Xorg.crouton.1.log", Time: Thu Feb 26 14:05:14 2015
(++) Using config file: "/etc/X11/xorg-dummy.conf"
(==) Using system config directory "/usr/share/X11/xorg.conf.d"
Parse error on line 36 of section ServerFlags in file /etc/X11/xorg-dummy.conf
        "/Option" is not a valid keyword in this section.
(EE) Fatal server error:
(EE) no screens found(EE) 
(EE) Please consult the The X.Org Foundation support at http://wiki.x.org for help. 
(EE) Please also check the log file at "/tmp/Xorg.crouton.1.log" for additional information.
(EE) 
(EE) Server terminated with error (1). Closing log file.
/usr/bin/xinit: giving up
/usr/bin/xinit: unable to connect to X server: Connection refused
/usr/bin/xinit: server error

**Comment by dnschneid** [comment]
Argh, my bad.  I fixed the original command, but before you run it again (above), run this to clear that messed up line out:

```
sudo enter-chroot -u root sed -i '/AutoAddGPU/d'
```

**Comment by Watney** [comment]
Just to confirm that I'm following correctly: 

1. ``sudo enter-chroot -u root sed -i '/AutoAddGPU/d' /etc/crouton/xorg-dummy.conf``
2. ``sudo enter-chroot -u root sed -i '/AutoAddDevices/aOption "AutoAddGPU" "false"' /etc/crouton/xorg-dummy.conf``
3. ``sudo enter-chroot exec env XMETHOD=xiwi startxfce4``

(EE) Segmentation fault at address 0x0
(EE) 
Fatal server error:
(EE) Caught signal 11 (Segmentation fault). Server aborting
(EE)

**Comment by dnschneid** [comment]
Yeah, that's right.  Could you now post the contents of /tmp/Xorg.crouton.1.log?

**Comment by Watney** [comment]
``chronos@localhost / $ cat /tmp/Xorg.crouton.1.log``
[  4410.323] _XSERVTransmkdir: Owner of /tmp/.X11-unix should be set to root
[  4410.325] 
X.Org X Server 1.15.1
Release Date: 2014-04-13
[  4410.325] X Protocol Version 11, Revision 0
[  4410.325] Build Operating System: Linux 3.2.0-76-generic x86_64 Ubuntu
[  4410.325] Current Operating System: Linux localhost 3.10.18 #1 SMP Tue Feb 24 17:56:23 PST 2015 x86_64
[  4410.325] Kernel command line: cros_secure console= loglevel=7 init=/sbin/init cros_secure oops=panic panic=-1 root=/dev/dm-0 rootwait ro dm_verity.error_behavior=3 dm_verity.max_bios=-1 dm_verity.dev_wait=1 dm="1 vroot none ro 1,0 2506752 verity payload=PARTUUID=59135242-6541-924e-85f8-cc30ec6fb713/PARTNROFF=1 hashtree=PARTUUID=59135242-6541-924e-85f8-cc30ec6fb713/PARTNROFF=1 hashstart=2506752 alg=sha1 root_hexdigest=3a9eb9759546c179f627e49977d8e660b78fe6ee salt=f5c31fd0f5dd4d87eb49564260af82fbf9ddb12d1fae9a7f439a7a919d590746" noinitrd vt.global_cursor_default=0 kern_guid=59135242-6541-924e-85f8-cc30ec6fb713 add_efi_memmap boot=local noresume noswap i915.modeset=1 tpm_tis.force=1 tpm_tis.interrupts=0 nmi_watchdog=panic,lapic  
[  4410.326] Build Date: 12 February 2015  02:49:29PM
[  4410.326] xorg-server 2:1.15.1-0ubuntu2.7 (For technical support please see http://www.ubuntu.com/support) 
[  4410.326] Current version of pixman: 0.30.2
[  4410.326]    Before reporting problems, check http://wiki.x.org
        to make sure that you have the latest version.
[  4410.326] Markers: (--) probed, (**) from config file, (==) default setting,
        (++) from command line, (!!) notice, (II) informational,
        (WW) warning, (EE) error, (NI) not implemented, (??) unknown.
[  4410.327] (++) Log file: "/tmp/Xorg.crouton.1.log", Time: Thu Feb 26 14:22:37 2015
[  4410.328] (++) Using config file: "/etc/X11/xorg-dummy.conf"
[  4410.328] (==) Using system config directory "/usr/share/X11/xorg.conf.d"
[  4410.329] (==) ServerLayout "Layout0"
[  4410.329] (**) |-->Screen "Screen0" (0)
[  4410.329] (**) |   |-->Monitor "Monitor0"
[  4410.330] (**) |   |-->Device "Card0"
[  4410.330] (**) Option "DontVTSwitch" "true"
[  4410.330] (**) Option "AllowMouseOpenFail" "true"
[  4410.330] (**) Option "AutoAddDevices" "false"
[  4410.330] (**) Option "AutoEnableDevices" "false"
[  4410.330] (**) Option "AutoAddGPU" "false"
[  4410.330] (**) Not automatically adding devices
[  4410.330] (**) Not automatically enabling devices
[  4410.330] (**) Not automatically adding GPU devices
[  4410.331] (WW) The directory "/usr/share/fonts/X11/cyrillic" does not exist.
[  4410.331]    Entry deleted from font path.
[  4410.331] (WW) The directory "/usr/share/fonts/X11/100dpi/" does not exist.
[  4410.331]    Entry deleted from font path.
[  4410.331] (WW) The directory "/usr/share/fonts/X11/75dpi/" does not exist.
[  4410.331]    Entry deleted from font path.
[  4410.331] (WW) The directory "/usr/share/fonts/X11/100dpi" does not exist.
[  4410.331]    Entry deleted from font path.
[  4410.331] (WW) The directory "/usr/share/fonts/X11/75dpi" does not exist.
[  4410.331]    Entry deleted from font path.
[  4410.331] (==) FontPath set to:
        /usr/share/fonts/X11/misc,
        /usr/share/fonts/X11/Type1,
        built-ins
[  4410.331] (==) ModulePath set to "/usr/lib/x86_64-linux-gnu/xorg/extra-modules,/usr/lib/xorg/extra-modules,/usr/lib/xorg/modules"
[  4410.331] (==) |-->Input Device "<default pointer>"
[  4410.331] (==) |-->Input Device "<default keyboard>"
[  4410.331] (==) The core pointer device wasn't specified explicitly in the layout.
        Using the default mouse configuration.
[  4410.331] (==) The core keyboard device wasn't specified explicitly in the layout.
        Using the default keyboard configuration.
[  4410.331] (II) Loader magic: 0x7fad81c78d40
[  4410.331] (II) Module ABI versions:
[  4410.331]    X.Org ANSI C Emulation: 0.4
[  4410.331]    X.Org Video Driver: 15.0
[  4410.331]    X.Org XInput driver : 20.0
[  4410.331]    X.Org Server Extension : 8.0
[  4410.332] (II) xfree86: Adding drm device (/dev/dri/card0)
[  4410.333] (II) xfree86: Adding drm device (/dev/dri/card1)
[  4410.334] (EE) 
[  4410.334] (EE) Backtrace:
[  4410.334] (EE) 0: /usr/bin/Xorg (xorg_backtrace+0x48) [0x7fad819f4848]
[  4410.335] (EE) 1: /usr/bin/Xorg (0x7fad8184b000+0x1ad539) [0x7fad819f8539]
[  4410.335] (EE) 2: /lib/x86_64-linux-gnu/libpthread.so.0 (0x7fad80948000+0x10340) [0x7fad80958340]
[  4410.335] (EE) 3: /usr/bin/Xorg (0x7fad8184b000+0xb57a6) [0x7fad819007a6]
[  4410.335] (EE) 4: /usr/bin/Xorg (xf86BusProbe+0x9) [0x7fad818d4099]
[  4410.336] (EE) 5: /usr/bin/Xorg (InitOutput+0x74d) [0x7fad818e26fd]
[  4410.336] (EE) 6: /usr/bin/Xorg (0x7fad8184b000+0x59bab) [0x7fad818a4bab]
[  4410.337] (EE) 7: /lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main+0xf5) [0x7fad7f388ec5]
[  4410.337] (EE) 8: /usr/bin/Xorg (0x7fad8184b000+0x451ee) [0x7fad818901ee]
[  4410.337] (EE) 
[  4410.338] (EE) Segmentation fault at address 0x0
[  4410.338] (EE) 
Fatal server error:
[  4410.338] (EE) Caught signal 11 (Segmentation fault). Server aborting
[  4410.339] (EE) 
[  4410.339] (EE) 
Please consult the The X.Org Foundation support 
         at http://wiki.x.org
 for help. 
[  4410.339] (EE) Please also check the log file at "/tmp/Xorg.crouton.1.log" for additional information.
[  4410.340] (EE)

**Comment by dnschneid** [comment]
Ok, it's still trying to get at /dev/dri for some reason...

**Comment by dnschneid** [comment]
Technical findings so far:

* dev recently added a vgem driver for various optimizations down the road
* the vgem driver doesn't have a pci bus id for obvious reasons
* libdrm2 returns NULL if the dri device doesn't implement DRM_IOCTL_GET_UNIQUE
* the X server in Ubuntu currently happily stores the NULL pointer and borks on it later
* I can't find a way to prevent scanning other than disabling udev support (which will break input devices).  It may be worth doing this on xiwi.
* LD_PRELOADing drmGetBusid to return "" instead of NULL might work, but might also break calls that properly handle a NULL pointer.  I'm trying this next.
* I'm also chatting with Chromium devs to check sanity and see if it's a bug on Chromium OS's side.

Note that this breaks BOTH xiwi and xorg on dev channel.

**Comment by Watney** [comment]
@dnschneid nothing like having someone ~~sabotage~~ bork your work :)

**Comment by dnschneid** [comment]
OK, drmGetBusid isn't even getting called, so none of that is the issue.  X11 is simply not setting the bus id because it's not a pci device, and then stupidly tripping over that later.

**Comment by dnschneid** [comment]
OK, try the latest crouton.  xorg should work.  xiwi doesn't yet.

**Comment by Watney** [comment]
@dnschneid **success!!** *you rock*

**Comment by dnschneid** [comment]
OK, xiwi should be running again as well.

**Comment by dwploc** [comment]
Everything starts fine but eventually this happens:

croutoncycle next|prev|cros|list|#
Cycles through running graphical chroots.
    next: switch to the next display
    prev: switch to the previous display
    cros: switch directly to Chromium OS
    list: list all of the active displays and the associated chroot name
    #: switch to the nth item in the list, zero-indexed
    :#: switch directly to the chroot owning the specified display number
^C
Session terminated, terminating shell.../usr/bin/xinit: connection to X server lost

waiting for X server to shut down Hangup
Hangup
Hangup
Running exit commands...

(xfwm4:849): libxfce4ui-WARNING **: ICE I/O Error

(xfwm4:849): libxfce4ui-WARNING **: Disconnected from session manager.

(xfsettingsd:916): libxfce4ui-WARNING **: ICE I/O Error

(xfsettingsd:916): libxfce4ui-WARNING **: Disconnected from session manager.
xfwm4: Fatal IO error 11 (Resource temporarily unavailable) on X server :1.0.

(xfce4-volumed:923): xfce4-volumed-CRITICAL **: xvd_context_state_callback: The connection failed or was disconnected, is PulseAudio Daemon running?
xfdesktop: Fatal IO error 11 (Resource temporarily unavailable) on X server :1.0.
xfsettingsd: Fatal IO error 11 (Resource temporarily unavailable) on X server :1.0.
xfce4-volumed: Fatal IO error 11 (Resource temporarily unavailable) on X server :1.0.
(EE) Server terminated successfully (0). Closing log file.

/usr/bin/xinit: unexpected signal 2
 ...terminated.
Unmounting /mnt/stateful_partition/crouton/chroots/tab...
Sending SIGTERM to processes under /mnt/stateful_partition/crouton/chroots/tab...
chronos@localhost / $

**Comment by dnschneid** [comment]
That's harmless.  It'll go away once xiwi-app is merged.

**Comment by dwploc** [comment]
Thx, you da man !!!  I no longer panic when one of these issues raises its head.

**Comment by superfawkes** [comment]
Sorry, a newbie here... in mid-March, I used the most obvious crouton install to get the xfce/debian-sid - which seems to be running under xephyr (I have an Acer CB13 Tegra K1). I've tried experimenting with a few other unsupported chroot configurations, without much success. I'd be glad to help test patches.

Is there a way to get the crouton script to pull patches that haven't been merged to the main branch?
